azavea/echo-locator
# Add password-less authentication flow for house seekers

## Overview

This PR gives house seekers the ability to log in to the app as an alternative to using it anonymously. It does so using a passwordless login flow in which the user enters their email and clicks sign in, receives an email with a sign in and link, and upon clicking it, is logged in and their profile loaded. The relationship between Django and React during this flow is illustrated below.

![Django Magic Link Auth (2)](https://user-images.githubusercontent.com/77936689/166920587-9ad10733-1b30-4687-9ff3-59a7a770a222.png)

### Notes

**Fetch userProfile**: As seen in the above picture, this login system takes advantage of the existing `authenticator` component (previously named `anonymous-authenticator`) which renders either the sign in page or the application depending on whether a `userProfile` exists. On render, `authenticator` checks for the unique case in which the user has an `auth_token` cookie but not a user profile or an error `loginMessage` as generated on failed login (see more on that below). If this is the case, it then requests the user's profile from the backend and either saves the profile to state or shows the error message. I chose to put it in render here, as opposed to in `handleAuthChange`, because this part of the authentication needs to happen without interaction on the frontend, but rather as a response to the backend's redirect to the frontend. The check for `auth_token`, `userProfile`, and `loginMessage` should be enough to avoid infinite rendering.

**loginMessage:** As mentioned above, this PR adds a state-wide `loginMessage` that is displayed on the sign in page first following submission of an email, and potentially again if the user attempting to sign in does not have a `userProfile`. This should currently only be necessary when a Counselor or admin who has not made themselves an account attempts to log in, but may be useful when adding user sign up in issue #465. 

**Failed Test:** This PR uses Django Rest Framework's Token Authentication, which added a new `Token` table that changed the `pk`s of the entries in `auth.permissions` as used by `src/backend/users/fixtures/group_permissions.json` to set the Counselor group's permissions, resulting in a failed test. This PR includes an edit to that file to allow Counselors to view and change Destinations, as well as Users and their profiles, and updates the `pk`s in that fixture to reflect the new ones following the addition of `Token`.

 **Dummy Data:** In order to successfully bring each profile from the backend into the frontend, the data needed to match the `AccountProfile` type. As a result, `GetUserProfile` view includes a repackaging of data retrieved from the `UserProfile` and `Destination` tables to match AccountProfile. As described in PR [457](https://github.com/azavea/echo-locator/pull/457), Destination should but does not currently include longitude/latitude data. This PR uses dummy data when sending Destinations to the frontend, to be replaced with real data following the merge of 457.

**Why 8085?:** Try as I might, I could not get the `ObtainToken` view to work via port 9966. As a result, the magic link directs to 8085 and everything past that occurs on 8085 to avoid CSRF issues.


## Testing Instructions

 * Run `./scripts/update` to migrate the new `Token` table and to apply the new Counselor group permissions
 * Start the server and navigate to **localhost:9966**. 
 * First, check that signing in with an invalid email address fails silently. Enter an email that you _do not_ have an account associated with and verify 1) you still see the success message on the sign in page, 2) you do not see an "email" in your terminal, and 3) that you do not see any 500 errors in your console or network tab of the dev tools.
 * Next, check that signing in as a counselor/admin displays the correct error message. Enter an email for which you _do_ have a user but _not_ a user profile and click "Sign in".
 * You should see an "email" in your terminal. Copy and paste the link into your browser. It should display a message on the screen explaining that you do not have a profile.
 * Check that login works for users with an associated profile. Enter an email for which you have an associated user and userProfile and click "Sign in".
 * You should see an "email" in your terminal. Copy and paste the link into your browser. It should take you into the app. 
 * Verify that you see the account's addresses in the dropdown menu of the dock (as a reminder, the routing is currently using dummy lon/lat data for each address, but you should still be able to see the proper addresses and labels in the dock).
 * Finally, run `./scripts/test` and verify that the fixture fix worked as expected.


Resolves #442
